Description: Allow for extra space in syslog parsing
Bug-Debian: https://bugs.debian.org/330974
Author: Stephen Gran <sgran@debian.org>

--- ccze-0.2.1.orig/src/mod_syslog.c
+++ ccze-0.2.1/src/mod_syslog.c
@@ -104,7 +104,7 @@
   int errptr;
 
   reg_syslog = pcre_compile ("^(\\S*\\s{1,2}\\d{1,2}\\s\\d\\d:\\d\\d:\\d\\d)"
-			     "\\s(\\S+)\\s((\\S+:?)\\s(.*))$", 0, &error,
+			     "\\s(\\S+)\\s+((\\S+:?)\\s(.*))$", 0, &error,
 			     &errptr, NULL);
   hints_syslog = pcre_study (reg_syslog, 0, &error);
 }
